onnxruntime/contrib_ops/cpu/quantization/qlinear_softmax.cc

```diff
@@ -0,0 +1,260 @@
+    QlinearBuildLookupTableUint32(fixed_lookup_table_.data(), X_scale, channels);
+  }
+}
+
+template <typename T>
+QLinearSoftmax<T>::QLinearSoftmax(const OpKernelInfo& info)
+    : OpKernel(info) {
+  const auto& node = info.node();
+  opset_ = node.SinceVersion();
+  int64_t axis = -1;
+  Status status = info.GetAttr<int64_t>("axis", &axis);
+  if (status.IsOK()) {
+    axis_ = gsl::narrow_cast<int>(axis);
+  } else {
+    if (opset_ < 13) {
+      axis_ = 1;  // opset-12 and below, the default axis value is 1
+    } else {
+      axis_ = -1;  // opset-13, the default axis value is -1
+    }
+  }
+  auto input_defs = node.InputDefs();
+  const auto& x_shape = input_defs[0]->Shape();
+  size_t rank = x_shape->dim_size();
+  if (rank == 0) {
+    return;
+  }
+  if (axis_ < 0) {
+    axis_ = static_cast<int>(HandleNegativeAxis(axis_, int64_t(rank)));
+  }
+  uint32_t channels = gsl::narrow_cast<uint32_t>(x_shape->dim(axis_).dim_value());
+  if (opset_ < 13) {
+    for (size_t i = axis_; i < rank; i++) {
+      channels *= gsl::narrow_cast<uint32_t>(x_shape->dim(i).dim_value());
+    }
+  }
+  ORT_ENFORCE(channels>0, "invalid channels for softmax");
+  this->BuildLookupTableIfFixed(info, channels);
+}
+
+// compute method of Softmax
+template <typename T>
+Status QLinearSoftmax<T>::Compute(OpKernelContext* ctx) const {
+  const auto* X = ctx->Input<Tensor>(0);
+  const auto& X_shape = X->Shape();
+  auto* Y = ctx->Output(0, X_shape);
+
+  // edge case. one or more dims with value of 0. nothing to do
+  if (X_shape.Size() == 0) {
+    return Status::OK();
+  }
+  concurrency::ThreadPool* thread_pool = ctx->GetOperatorThreadPool();
+  size_t D = X_shape[axis_];
+  if (opset_ < 13) {
+    D = X_shape.SizeFromDimension(axis_);
+  }
+  const uint32_t* lookup_table = GetLookupTable(ctx, D);
+
+  if (opset_ < 13) {
+    return ComputeImpl(*X, *Y, thread_pool, lookup_table);
+  } else {
+    return ComputeImplOpset13(ctx, *X, *Y, thread_pool, lookup_table);
+  }
+}
+
+common::Status QlinearSoftmaxCPU(size_t N,
+                                 size_t D,
+                                 const uint8_t* x_data,
+                                 uint8_t* y_data,
+                                 const uint32_t* lookup_table,
+                                 onnxruntime::concurrency::ThreadPool* thread_pool) {
+  using onnxruntime::TensorOpCost;
+  using onnxruntime::concurrency::ThreadPool;
+  ThreadPool::TryParallelFor(
+      thread_pool, N, TensorOpCost{1.0, 1.0, 1.0},
+      [x_data, y_data, D, &lookup_table](std::ptrdiff_t first, std::ptrdiff_t last) {
+        int n = last - first;
+        if (n == 0) {
+          return;
+        }
+        const uint8_t* x_t = x_data + first * D;
+        uint8_t* y_t = y_data + first * D;
+        for (; first < last; first++) {
+          // reduceMax
+          uint8_t xmax = *std::max_element(x_t, x_t + D);
+          const size_t adjustment = xmax ^ 255;
```

Comment:
>adjustment [](http://example.com/codeflow?start=23&length=10)

is the adjustment necessary?

Reply:
Yes,  The LookupTable is build by exp(x-255), we assuming 255 is always the max_value in input tensor.
We can get  the real max value in the middle of computation, so we can add a shift (255-x_max) for each number, but we just adjust lookup table for convenience .